flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBIncrementalRestoreOperation.java

```diff
@@ -0,0 +1,507 @@
+		ColumnFamilyOptions columnOptions,
+		RocksDBNativeMetricOptions nativeMetricOptions,
+		MetricGroup metricGroup,
+		@Nonnull Collection<KeyedStateHandle> restoreStateHandles) {
+		super(keyGroupRange,
+			keyGroupPrefixBytes,
+			numberOfTransferringThreads,
+			cancelStreamRegistry,
+			userCodeClassLoader,
+			kvStateInformation,
+			keySerializerProvider,
+			instanceBasePath,
+			instanceRocksDBPath,
+			dbOptions,
+			columnOptions,
+			nativeMetricOptions,
+			metricGroup,
+			restoreStateHandles);
+		this.operatorIdentifier = operatorIdentifier;
+		this.restoredSstFiles = new TreeMap<>();
+		this.lastCompletedCheckpointId = -1L;
+		this.backendUID = UUID.randomUUID();
+	}
+
+	/**
+	 * Root method that branches for different implementations of {@link KeyedStateHandle}.
+	 */
+	@Override
+	public RocksDBRestoreResult restore() throws Exception {
+
+		if (restoreStateHandles == null || restoreStateHandles.isEmpty()) {
+			return null;
+		}
+
+		final KeyedStateHandle theFirstStateHandle = restoreStateHandles.iterator().next();
+
+		boolean isRescaling = (restoreStateHandles.size() > 1 ||
+			!Objects.equals(theFirstStateHandle.getKeyGroupRange(), keyGroupRange));
+
+		if (isRescaling) {
+			restoreWithRescaling(restoreStateHandles);
+		} else {
+			restoreWithoutRescaling(theFirstStateHandle);
+		}
+		return new RocksDBRestoreResult(this.db, defaultColumnFamilyHandle,
+			nativeMetricMonitor, lastCompletedCheckpointId, backendUID, restoredSstFiles);
+	}
+
+	/**
+	 * Recovery from a single remote incremental state without rescaling.
+	 */
+	void restoreWithoutRescaling(KeyedStateHandle rawStateHandle) throws Exception {
+		// 1. Prepare for restore without rescaling
+		Path temporaryRestoreInstancePath = new Path(
+			instanceBasePath.getAbsolutePath(),
+			UUID.randomUUID().toString()); // used as restore source for IncrementalKeyedStateHandle
+		RocksDBIncrementalRestorePrepareResult prepareResult = prepareFiles(rawStateHandle, temporaryRestoreInstancePath);
+		Path restoreSourcePath = prepareResult.getLocalKeyedStateHandle().getDirectoryStateHandle().getDirectory();
+		if (rawStateHandle instanceof IncrementalKeyedStateHandle) {
+			backendUID = ((IncrementalKeyedStateHandle) rawStateHandle).getBackendIdentifier();
+		} else {
+			backendUID = ((IncrementalLocalKeyedStateHandle) rawStateHandle).getBackendIdentifier();
+		}
+		LOG.debug("Restoring keyed backend uid in operator {} from incremental snapshot to {}.",
+			this.operatorIdentifier, this.backendUID);
+		if (!instanceRocksDBPath.mkdirs()) {
+			String errMsg = "Could not create RocksDB data directory: " + instanceBasePath.getAbsolutePath();
+			LOG.error(errMsg);
+			throw new IOException(errMsg);
+		}
+		try {
+			restoreInstanceDirectoryFromPath(restoreSourcePath, dbPath);
+		} finally {
+			FileSystem restoreFileSystem = restoreSourcePath.getFileSystem();
+			if (restoreFileSystem.exists(restoreSourcePath)) {
+				restoreFileSystem.delete(restoreSourcePath, true);
```

Comment:
If we just to `try-finally`, an exception that happens during IO in the `finally` block can cause the loss of the original exception. There are more cases of this problem in the class. One way is that you could use `ExceptionUtils.firstOrSuppressed` to preserve original exceptions.